### DISCUSSION TITLE
Looking for Interceptor on Publish to handle Backpressure

### DISCUSSION THREAD

**Discussion by ironydips** [uNetworking/uWebSockets.js#897] [Q&A] [answered]
Trying my hands on uwebsockets.js first time. Super fine experience but stuck badly now.

We have 1000 clients connected to nodejs uws. And we have to send same data (~350KB) to all clients as they subscribed by default on `open`. For this we are calling `publish` but 30% of the msgs are getting lost due to Backpressure.

How do we handle this as we do not see any method to interfere after `publish` is called, data directly goes to Client or fails.

Some Code Snippet is highly appreciated.

Also, things are pretty smooth when we run on local machine, but on ec2 server Backpressure issue becomes 100 times more

**Comment by e3dio** [comment, marked as answer]
Adjust your settings: [maxBackpressure](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocketBehavior.html#maxBackpressure), [closeOnBackpressureLimit](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocketBehavior.html#closeOnBackpressureLimit)

And minimize your data size. How often are you publishing 350kb to 1000 clients? Are you sure that is the smallest possible size? What kind of data/format is taking up 350kb on every publish?

**Comment by ironydips** [comment]
Though I am reading a lot about maxBackpressure, can not get full hold of it.

- Data Transfer Frequency is every 1 min.
- Clients will gradually increase to 10,000 once we are fully confident of msgs not being dropped as our data is related to Stock Market and timely updates are super crucial for financial markets
- we are stringifying data using JSON currently, exploring other ways of fast mechanism & to reduce size.

```
const app = uWS.App()
app.ws('/*', {
     compression: uWS.SHARED_COMPRESSOR,
     maxPayloadLength: 16 * 1024,
     idleTimeout: 30,
     maxBackpressure: 1024 * 100
})

```

**Reply by uNetworkingAB** [reply]
The default setting is to drop new messages that would cause more than maxBackpressure bytes of backpressure. If you don't want to drop messages you either have to increase maxBackpressure to 1024 * 1024 or something like that, but messages are going to be dropped if they need to be. The other alternative is to enable closeOnBackpressureLimit but this just makes the whole connection drop on too much backpressure. There are no other options, you either drop or close, or you disable it by setting maxBackpressure to 0 but then any slow receiver can down your entire server.

**Reply by ironydips** [reply]
1) so right way should be manage data to be sent in better way than json.stringify to reduce size?

any suggestions on that front pls?


2) also is there any settings i might have to add on aws ec2 instance because locally everything is fine, just that when same thing connects to server... not even 5 connections sustain properly without dropping.

i am counting so much on this library... thank you so much for your efforts in developing & maintaining this.

3) One more thing. I was unable to find any example of using streams in uws.js to send data either through ws.send or ws.publish... not sure if that would help improve the backpressure issue, can you pls point to some examples?

**Reply by e3dio** [reply]
350kb is huge amount of data, you can reduce that by 10x or more easily with correct format 35kb, 10x less data sent = 10x less backpressure, this might solve your problem yes? Try this library https://github.com/e3dio/packBytes and open discussion there or here if you need help, include sample data if you want specific example

**Reply by ironydips** [reply]
> 350kb is huge amount of data, you can reduce that by 10x or more easily with correct format 35kb, 10x less data sent = 10x less backpressure, this might solve your problem yes? Try this library https://github.com/e3dio/packBytes and open discussion there or here if you need help, include sample data if you want specific example

thank you for this library, looks very fast. I see it is currently using `import` es6 version, can't we use `require` here as i have all other moduels using commonjs, any workaround? Also do we have some more complex examples using this library?

**Reply by uNetworkingAB** [reply]
My personal favourite opinion when it comes to market data is: don't expose JSON APIs - expose a connector library in JavaScript that is super easy to use, while internally using very efficient binary protocols. Then you get two benefits: quick adoption if your connector library is easy to use in just a handful of lines, and you never lock yourself to some arcane format that will eventually need compression to even reliably work.

**Reply by ironydips** [reply]
sure, will try this one too.. thank you

**Reply by ironydips** [reply]
> 350kb is huge amount of data, you can reduce that by 10x or more easily with correct format 35kb, 10x less data sent = 10x less backpressure, this might solve your problem yes? Try this library https://github.com/e3dio/packBytes and open discussion there or here if you need help, include sample data if you want specific example

@e3dio Need help. As per I have read in packBytes Doc, it is faster than AVSC, but when I am testing it then AVSC seems to outperform PackBytes by decent margin. I am Dead sure I am doing some mistake as I am new to it. Please help me find the issue here and optimise Schema of PackBytes

Below is the code snippet. I am putting only 10% of my actual data in the file for better understanding (I know it's big data for an example, but it's needed to test.)

Here are my results for this small subset. AVSC > PackBytes > JSON

```
JSON String length 121219

PackBytes Length 35147

AV Buffer Length 30600

```

Also, if I only GZIP raw data output is 18834

I am confused what will be the best, PackByte, GZIP or AVSC.

```
const { bool, bits, string, array, float, blob, schemas, PackBytes } = require('./lib/packBytes');
const avro = require('avsc'); 

// Bigger data is In the FILE
[demo-avsc-packbytes.zip](https://github.com/uNetworking/uWebSockets.js/files/11396517/demo-avsc-packbytes.zip)

 
const optData = {
    "type": "n_opt",
    "D": 230504,
    "T": 1434,
    "I": 1,
    data: [
        {
            "E": 230504,
            "S": [
                {
                    "D": 230504,
                    "T": 915,
                    "C": 134.45,
                    "V": 124250,
                    "St": 17950,
                    "OI": 629950,
                    "Ins": 0
                }
             ],
            "M": [
                {
                    "D": 230504,
                    "T": 1434,
                    "C": 254.05,
                    "V": 17100,
                    "St": 17950,
                    "OI": 492100,
                    "Ins": 0
                }
            ]
         }
    ]
}

const marketDataSchema = {
    D: bits(20),
    T: bits(11),
    C: float(32),
    V: bits(22),
    St: bits(20),
    OI: bits(32),
    Ins: bits(1)
}

const optDataSchema = {
    E: bits(20),
    S: array(marketDataSchema),
    M: array(marketDataSchema)
}

// main schema
const optSchema = {
    type: string,
    I: bits(2),
    D: bits(20),
    T: bits(11),
    data: array(optDataSchema)
 };

const encoder = new PackBytes(optSchema);

const buf = encoder.encode(optData);
console.log('JSON String length', JSON.stringify(optData).length)
console.log(`PackBytes Length`, buf.length)
const res = encoder.decode(buf)
//  console.log(res)

 const AVtype = avro.Type.forValue({
    "type": "n_opt",
    "D": 230504,
    "T": 1434,
    "I": 1,
    "data": [optData.data[0]]
})

const avBuffer = AVtype.toBuffer(optData)
console.log('AV Buffer Length', avBuffer.length)
const avRes = AVtype.fromBuffer(avBuffer)
```

**Reply by e3dio** [reply]
You were close to getting best optimized encoding for your data, here is what I get after optimize for your case:

. | Bytes
--- | ---
PackBytes | 14,839
Avsc | 18,843
Json | 73,648

99% of your data size is `marketDataSchema` so you want to optimize that as much as possible. I take the minimum and maximum value for each field to get the range, find bits required to represent the range and use that for the field. Also I take your float(32) field 'C' and multiply by 100 to make it integer which uses 19 bits instead of 32, this brings pack size from 16bytes/item to 14bytes/item
```js
const marketDataSchema = {
   D: bits(8),
   T: bits(10),
   C: bits(19),
   V: bits(23),
   St: bits(13), 
   OI: bits(25),
   Ins: bits(1),
}
```
PackBytes runs algorithm to efficiently pack each `marketDataSchema` item into 14 bytes, 14 bytes * 1057 items in your dataset = 14,798 bytes, and the remainder of the 14,839 bytes comes from array sizes and those top level fields.

Your results you posted were slightly different than the results I got running the data you provided in that zip file, you should check to make sure you are running the same data if you want to directly compare results

**Reply by ironydips** [reply]
thanks a ton @e3dio for taking so much effort to help me out.

firstly, this 'C' logic of making it integer is just Wow. Simple, Elegant, & space saving.

(2) I think D: bits(8) will not be right as value it holds is a date in format 'YYMMDD', its integer with 20 bits holding. Is my understanding correct?

(3) Yes I realise I have posted results for larger data set, Sorry for confusion. But you got it right.

Most Important part: I tried gzipping raw data at last (my bad) and saw that it has BEST minimal output amongst every thing. It doesn't ask us to add any schema. I am little confused, should we actually use gzip on raw data and don't use any other library for compression. Am I missing anything in this simplistic logic. Because I am getting a feel if gzip can perform so better than why take extra pain with schema-based-libraries. (just for my better understanding)

Thanks again :)

**Reply by e3dio** [reply]
You have many of repeating values in your dataset, I don't know if this is normal for your data or it's because you copy/pasted to create demo data. Gzip can take advantage of repeating strings to make it smaller. Gzip+Json produced 12,033 bytes compared to 14,839 for PackBytes which is not bad, but it also took 22x more CPU time. You could use it if it works for you, depends on the case. If you have many repeating values you could re-code your schema to take advantage of that

**Reply by ironydips** [reply]
that clears all the compression doubts. also, this is the production data I shared and since many values are repeated (normal for our use case) I now see why gzip worked just fine.

since data is now reduced to under 40kb now (thanks to you), I shall now monitor my original problem of BackPressure & get back with my findings/further queries.

not sure if this is right forum to ask this or not, I am dedicatedly using aws ec2 server as r6.xlarge for websocket purpose, can you pls suggest if such instances are optimal for sending ~40kb of data every minute to roughly 3000 users. Or what kind of configuration should we look for in server instance?

**Reply by e3dio** [reply]
Ec2 has at least 5 Gb per second network out, your data is 40kB * 3000 = .12GB = .96Gb per minute so it has more than enough out

**Reply by e3dio** [reply]
I Gzip'd the PackBytes encoding and it was 3x faster and 22% smaller than Gzip+Json, so if you really want the smallest size you know what to do ;)

. | Bytes
--- | ---
Gzip+PackBytes | 9,271
Gzip+Json | 12,033
PackBytes | 14,839
Avsc | 18,843
Json | 73,648

**Reply by ironydips** [reply]
I also just tried the same gzipping packbytes and was about to post here.. way to go. :)

just confused about the D: bits(8) rest all clear :)

**Reply by e3dio** [reply]
Max value for D was 230504 and min value was 230320 which means there are 184 possible values which requires 8 bits to represent. You say this represents a Date field, you can change this field to represent 'number of days since X date' and use however many bits are needed, convert back to Date after decode

**Comment by e3dio** [comment]
> Looking for Interceptor on Publish to handle Backpressure

Should also mention to answer original question, if you want to handle each websocket connection individually depending on backpressure level, loop through each websocket and check `ws.getBufferedAmount()` for backpressure, and send custom messages  with `ws.send()` instead of `app.publish()`. But most important thing is get data size as small as possible, then you have less backpressure
